File: leetcode/需要二刷的题/110 平衡二叉树/official.c

int height(struct TreeNode* root) {
    if (root == NULL) {
        return 0;
    }
    int leftHeight = height(root->left);
    int rightHeight = height(root->right);
    if (leftHeight == -1 || rightHeight == -1 || fabs(leftHeight - rightHeight) > 1) {
        return -1;
    } else {
        return fmax(leftHeight, rightHeight) + 1;
    }
}

bool isBalanced(struct TreeNode* root) {
    return height(root) >= 0;
}

作者：LeetCode-Solution
链接：https://leetcode-cn.com/problems/balanced-binary-tree/solution/ping-heng-er-cha-shu-by-leetcode-solution/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

//大概能理解思路，就是我递归不会写呗……
//思路是类似的，就是避免height重复计算，官方用到的递归思想就是在求高度的时候同时判断是否是平衡二叉树，如果不平衡，给一个flag=-1，标记一下然后一层层的退出递归
//height求的是单个节点的高度，肯定大于0，而如果遇见不平衡二叉树的情况，则会返回-1，因此主函数的height(root) != -1 也是可以的
//后序遍历思想，就是先左子节点，然后右子节点，然后再自己，这个过程

File: leetcode/需要二刷的题/671 二叉树中第二小的节点/official.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findSecondMinimumValue(root *TreeNode) int {
	min := root.Val
	ans := -1
	var dfs func(*TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil || ans != -1 && node.Val >= ans {
			return
		}
		if node.Val > min {
			ans = node.Val
		}
		dfs(node.Left)
		dfs(node.Right)
	}
	dfs(root)
	return ans
}

//这是真的一点不会go啊……函数还能像13行这么定义吗？
//当然，也是真的一点不会DFS……DFS搜索的思想还没有深入我心，这还有空摸鱼？还不多多敲代码！